### PR TITLE
Add ablity to see user's top score in beatmap overlay

### PR DESCRIPTION
**Pull request by EVAST9919**
Screenshot_1

### CODE REVIEW

**Comment by peppy** [comment]
I think we ended up decidingto keep the # prefix for ranks, because it feels weird otherwise.

**Comment by EVAST9919** [comment]
So should I just delete the "#"?

**Comment by peppy** [comment]
No it should be there. I think I checked this PR before your last commit, all good.

File: osu.Game/Overlays/BeatmapSet/Scores/ScoresContainer.cs [hunk 1]
```diff
@@ -97,6 +104,20 @@ public IReadOnlyList<ScoreInfo> Scores
             }
         }
 
+        private APILegacyUserTopScoreInfo userScore;
+
+        public APILegacyUserTopScoreInfo UserScore
```
Comment (peppy) [inline osu.Game/Overlays/BeatmapSet/Scores/ScoresContainer.cs]: This feels *really* awkward. Why is this here, and why does it cancel the request?

**Comment by peppy** [comment]
In the current web design, the grade is beneath the #rank. Is there a reason you didn't follow that layout?

**Comment by EVAST9919** [comment]
> In the current web design, the grade is beneath the #rank. Is there a reason you didn't follow that layout?

B
52319928-9093eb80-2a0f-11e9-9044-a5130cebce37
ut isn't that latest design?

**Comment by peppy** [comment]
Confirmed with @arflyte that it should be below for now (to allow for more horizontal space).

File: osu.Game/Overlays/BeatmapSet/Scores/ScoresContainer.cs (hunk 1 above)
Comment (EVAST9919) [reply osu.Game/Overlays/BeatmapSet/Scores/ScoresContainer.cs]: Fixed, should be better now

File: osu.Game/Overlays/BeatmapSet/Scores/ScoresContainer.cs [hunk 2]
```diff
@@ -44,100 +39,165 @@ public ScoresContainer()
...
-                beatmap = value;
+            topScoresContainer.Clear();
 
-                Scores = null;
+            scoreTable.Scores = scores?.Scores.Count > 1 ? scores.Scores : new List<APILegacyScoreInfo>();
+            scoreTable.FadeTo(scores?.Scores.Count > 1 ? 1 : 0);
 
-                if (beatmap?.OnlineBeatmapID.HasValue != true)
-                    return;
+            if (scores?.Scores.Any() == true)
```
Comment (VPeruS) [inline osu.Game/Overlays/BeatmapSet/Scores/ScoresContainer.cs]: You can remove "== true" part

Comment (iiSaLMaN) [reply osu.Game/Overlays/BeatmapSet/Scores/ScoresContainer.cs]: boolean == const boolean is redundant

Comment (EVAST9919) [reply osu.Game/Overlays/BeatmapSet/Scores/ScoresContainer.cs]: we are using a null check here, so it's required

Comment (iiSaLMaN) [reply osu.Game/Overlays/BeatmapSet/Scores/ScoresContainer.cs]: `scores?.Scores.Any() ?? false`

File: osu.Game/Overlays/BeatmapSet/Scores/ScoresContainer.cs [hunk 3]
```diff
@@ -44,100 +39,165 @@ public ScoresContainer()
...
-                api.Queue(getScoresRequest);
+                if (userScore != null && userScore.Position != 1)
+                    topScoresContainer.Add(new DrawableTopScore(userScore.Score, userScore.Position));
             }
+
+            Loading = false;
         }
 
-        private void updateDisplay()
+        private class ContentContainer : VisibilityContainer
```
Comment (peppy) [inline osu.Game/Overlays/BeatmapSet/Scores/ScoresContainer.cs]: i don't like the existence of this class. specifically all the custom maxHeight logic, and the fact that the ResizeHeightTo in UpdateAfterChildren is going to fight with the `PopIn`/`PopOut` (I can see the early return, but still ugly). We should not be adding local implementations like this all over the place.

I propose you remove all this custom logic and get a base implementation working, then we can discuss how to make the transitions appear better.

File: osu.Game/Overlays/BeatmapSet/Scores/ScoresContainer.cs [hunk 4]
```diff
@@ -44,100 +39,165 @@ public ScoresContainer()
...
                     Direction = FillDirection.Vertical,
-                    Spacing = new Vector2(0, spacing),
-                    Margin = new MarginPadding { Vertical = spacing },
+                    AutoSizeAxes = Axes.Y,
+                    RelativeSizeAxes = Axes.X,
                     Children = new Drawable[]
                     {
-                        topScore = new DrawableTopScore(),
-                        scoreTable = new ScoreTable
+                        loadingContainer = new LoadingContainer
```
Comment (peppy) [inline osu.Game/Overlays/BeatmapSet/Scores/ScoresContainer.cs]: code-wise and visually, this feels weird. it's pushing content down physically, which isn't what a loading spinner should do.

File: osu.Game/Overlays/BeatmapSet/Scores/TopScoreUserSection.cs [hunk 5]
```diff
@@ -28,7 +28,7 @@ public class TopScoreUserSection : CompositeDrawable
         private readonly SpriteText date;
         private readonly UpdateableFlag flag;
 
-        public TopScoreUserSection()
+        public TopScoreUserSection(int position)
```
Comment (peppy) [inline osu.Game/Overlays/BeatmapSet/Scores/TopScoreUserSection.cs]: this can't be in the `ctor` is you leave the `score` as a property. it makes no sense, as you are able to update one but not the other.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
What was the reason for moving the request logic into the more general overlay class? Seems like it would still fit better inside `ScoresContainer`, no?

**Comment by EVAST9919** [comment]
> What was the reason for moving the request logic into the more general overlay class? Seems like it would still fit better inside `ScoresContainer`, no?

Just to allow adding scores manually only without being confused by having the request inside of it. That makes more sense imo. What's the point of having ability to set scores manually then if we already have all the logic needed for getting scores without that implementstion? Just for testcases? So that's why I moved it outside.

**Comment by peppy** [comment]
I would argue that a better solution would be to only expose that "manual" setting via a custom test class. We generally avoid any kind of `public` exposure just for test purposes (use `protected` at most).

I'd highly recommend the request is still done internally here, because we are doing our best to keep responsibilities as local to classes as possible.

**Comment by EVAST9919** [comment]
Alright, I'll put it back

**Review by peppy — APPROVED** [review]
